From 539efde90bf166bcda0af47ca857a75a9d5b76ed Mon Sep 17 00:00:00 2001
From: Steve Dickson <SteveD@redhat.com>
Date: Thu, 24 Jul 2014 18:16:11 -0400
Subject: [fs] nfs: handle multiple reqs in nfs_wb_page_cancel

Message-id: <1406225773-16395-9-git-send-email-steved@redhat.com>
Patchwork-id: 86637
O-Subject: [RHEL7.1 PATCH 08/10] nfs: handle multiple reqs in nfs_wb_page_cancel
Bugzilla: 1120728
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>

From: Weston Andros Adamson <dros@primarydata.com>

Use nfs_lock_and_join_requests to merge all subrequests into the head request -
this cancels and dereferences all subrequests.

Signed-off-by: Weston Andros Adamson <dros@primarydata.com>
Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Upstream Id: 3e2170451e91327bfa8a82040fea78043847533a
Bugzilla: 1120728

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfs/write.c b/fs/nfs/write.c
index c9f7d32..461156e 100644
--- a/fs/nfs/write.c
+++ b/fs/nfs/write.c
@@ -1820,27 +1820,28 @@ int nfs_wb_page_cancel(struct inode *inode, struct page *page)
  struct nfs_page *req;
  int ret = 0;
 
- for (;;) {
-  wait_on_page_writeback(page);
-  req = nfs_page_find_head_request(page);
-  if (req == NULL)
-   break;
-  if (nfs_lock_request(req)) {
-   nfs_clear_request_commit(req);
-   nfs_inode_remove_request(req);
-   /*
-    * In case nfs_inode_remove_request has marked the
-    * page as being dirty
-    */
-   cancel_dirty_page(page, PAGE_CACHE_SIZE);
-   nfs_unlock_and_release_request(req);
-   break;
-  }
-  ret = nfs_wait_on_request(req);
-  nfs_release_request(req);
-  if (ret < 0)
-   break;
+ wait_on_page_writeback(page);
+
+ /* blocking call to cancel all requests and join to a single (head)
+  * request */
+ req = nfs_lock_and_join_requests(page, false);
+
+ if (IS_ERR(req)) {
+  ret = PTR_ERR(req);
+ } else if (req) {
+  /* all requests from this page have been cancelled by
+   * nfs_lock_and_join_requests, so just remove the head
+   * request from the inode / page_private pointer and
+   * release it */
+  nfs_inode_remove_request(req);
+  /*
+   * In case nfs_inode_remove_request has marked the
+   * page as being dirty
+   */
+  cancel_dirty_page(page, PAGE_CACHE_SIZE);
+  nfs_unlock_and_release_request(req);
  }
+
  return ret;
 }
 
-- 
1.7.1